effectful-core/src/Effectful/Internal/Unlift.hs

```diff
@@ -107,6 +221,35 @@ persistentConcUnliftIO cleanUp threads k es0 unEff = do
               newEntries `seq` pure (newEntries, es)
     unEff m es
 
+----------------------------------------
+-- Errors
+
+-- | An type used for errors occuring in unlifting functions.
+data UnliftError
+  = InvalidNumberOfThreads Int
+  | InvalidNumberOfUses Int
+  | InvalidUseOfSeqUnlift
+  | ExceededNumberOfThreads Int
+  | ExceededNumberOfUses Int
+  deriving Show
+
+instance Exception UnliftError where
+  displayException (InvalidNumberOfThreads threads) =
```

Comment:
I wonder if this shouldn't be defined as `show` :thinking:

Reply:
Yes, this is the pain point of this PR: We do want nice error messages but sadly `displayException` never took off. And `show` is expected to output some internal representation of the error - Haskell code, most of the time.
I thought that might be some compromise but I am not really happy with that myself. Doing that there won't be any surprises for users of the `Show` instance and users who want the messages can use `displayException`.

Reply:
Ok, let's add a haddock note to the `UnliftError` definition that pretty printing can be done via `displayException`, because I didn't even know it existed.

Reply:
I added a note.